### ISSUE TITLE
`wasmprinter`: unknown 0xfb subopcode: 0x0

### ISSUE THREAD

**Issue by oovm** [bytecodealliance/wasm-tools#1335]
- Input

```
// wasmprinter 0.2.75
wasmprinter::print_bytes
```

[structure-example.zip](https://github.com/bytecodealliance/wasm-tools/files/13658641/structure-example.zip)

- Error

```js
unknown 0xfb subopcode: 0x0 (at offset 0x4e)
```

- Expect

```wat
(module $structure-example
  (type $Point (;0;) (struct (field $x f32) (field $y f32)))
  (memory $memory (;0;) (export "memory") 2)
  (func $Point::construct (;0;) (param $a (;0;) f32) (result (ref struct))
    (local $self.x f32)
    (local $self.y f32)
    local.get $a
    local.set $self.x
    local.get $a
    local.set $self.y
    local.get $self.x
    local.get $self.y
    struct.new $Point
  )
  (func $_start (;1;) (export "_start") (result (ref struct))
    f32.const 3.1400001
    call $Point::construct
  )
)
```

**Comment by oovm** [comment]
It seems that the parameter parsing of `struct.new` is wrong.

![image](https://github.com/bytecodealliance/wasm-tools/assets/17541209/3be82956-4e5d-4508-a2e7-eac7e91f61fe)

**Comment by alexcrichton** [comment]
Thanks for the report! The implementation of the GC proposal for WebAssembly is underway in this repository and it's known to not be complete. In that sense this is expected since tooling isn't complete yet.

**Comment by alxgrk** [comment]
@alexcrichton I ran into this error as well and started working on implementing the struct and array related instructions for the wasmparser crate. Would you be interested in a contribution?

**Comment by alexcrichton** [comment]
That'd be awesome, yes! I'd recommend coordinating with @fitzgen who's also working in this area to ensure hopefully no overlap as well

**Comment by fitzgen** [comment]
@alxgrk I am currently working on enabling the GC spec tests. That involves uncommenting one of these lines that tell our test suite to ignore the GC tests

https://github.com/bytecodealliance/wasm-tools/blob/6544a14d247a0fc9828473b3d56b24b74f37d631/tests/roundtrip.rs#L148-L157

and then running the tests with

```
cargo test --test roundtrip gc
```

and inspecting the error messages to see what the test needs that we are missing and implementing that and keep running the tests until the only failures are that we don't have test expectations which can be solved by running

```
BLESS=1 cargo test --test roundtrip
```

Once the tests are passing again, I send up a PR.

I'm working my way down from the top of that list, so feel free to work up from the bottom of the list.

**Comment by alxgrk** [comment]
@fitzgen this is the part I have so far: https://github.com/bytecodealliance/wasm-tools/compare/main...alxgrk:wasm-tools:main

It includes more changes than only enabling the `struct.wast` spec, since I came from a specific wasm binary that I wanted to parse without errors (which worked so far), but I still hope that my current state is useful.

Also, I started working on the `array.wast` spec already, so if you still see fit, I could also commit my current implementation to see if it could serve a baseline for enabling the `array*.wast` tests.

My current commit probably can be refactored as well, especially looking at the `visit_*` implementations in `crates/wasmparser/src/validator/operators.rs`. Regarding the git history, if you would like to have smaller, more self-contained commits before actually considering to merge a PR, I could surely split this very large commit up into smaller ones.
